PriceCalculater/Calculater.cs

```diff
@@ -12,27 +12,53 @@ public Calculater(ITaxService taxService , IDiscountService discountService )
         decimal amount = (percentage / 100) * price;
-        return ApplyPrecision(amount);
+        return amount.ApplyPrecision();
     }
-    public decimal ApplyPrecision(decimal price)
+        private void FindProductDetails(Product product)
     {
-        return Math.Round(price, 2);
+        productPriceDetails = new ProductPriceDetails();
+        decimal discountBeforeTax = CalculateDiscountBefore(product);
+        decimal remaningPrice = product.Price - discountBeforeTax;
+        productPriceDetails.TaxAmount = Calculate(remaningPrice, _taxService.GetTaxPercentage());
+        decimal discountAfterTax = CalculateDiscountAfter(product, remaningPrice);
+        productPriceDetails.DiscountAmount = discountBeforeTax + discountAfterTax;
+        productPriceDetails.FinalPrice = product.Price + productPriceDetails.TaxAmount - productPriceDetails.DiscountAmount.ApplyPrecision();
     }
-     private void FindProductDetails(Product product)
+   
+    public decimal CalculateFinalPrice(Product product) {
+        FindProductDetails(product);
+        return productPriceDetails.FinalPrice;
+    }
+    public decimal CalculateDiscountBefore(Product product) {
+        List<Discount> discounts = GetDiscounts(product);
+        decimal remainingPrice = product.Price;
+        var result =discounts.Select(discounts=>discounts).
+                     Where(d=>d.Prcedence==Precednce.before);
+        decimal dicountBefore = 0;
+        foreach(var item in result)
+        {
+            decimal discount  = Calculate(remainingPrice, item.Value);
+            dicountBefore += discount;
+            remainingPrice -= discount;
+        }
+        return dicountBefore;
+    }
+    public decimal CalculateDiscountAfter(Product product ,decimal price)
     {
-         productPriceDetails = new ProductPriceDetails();
-        productPriceDetails.TaxAmount = Calculate(product.Price, _taxService.GetTaxPercentage());
-        List<Discount>  discounts = _discountService.GetDiscountPercentage(product);
-        foreach(Discount discount in discounts) { 
-            productPriceDetails.DiscountAmount+=Calculate(product.Price, discount.Value);
+        decimal discountAfter = 0;
+        List<Discount> discounts = GetDiscounts(product);
+        var result = discounts.Select(discounts => discounts).
+                     Where(d => d.Prcedence == Precednce.after);
+        foreach(var item in result)
+        {
+            discountAfter += Calculate(price, item.Value);
         }
-        productPriceDetails.FinalPrice = ApplyPrecision(product.Price + productPriceDetails.TaxAmount - productPriceDetails.DiscountAmount);
-        
+        return discountAfter;
     }
-    public decimal CalculatePrice(Product product) {
-        FindProductDetails(product);
-        return productPriceDetails.FinalPrice;
+    private List<Discount> GetDiscounts(Product product)
+    {
+        return _discountService.GetDiscountPercentage(product);
     }
 }
```

Comment:
3) folder of calculater : main changes in method 1) dont change the states 2)loop at all discounts "treat them same " 

GetDiscounts: extract discount